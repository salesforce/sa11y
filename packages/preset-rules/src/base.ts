/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { getA11yConfig, RuleInfo } from './rules';

export const baseRulesInfo: RuleInfo = new Map(
    Object.entries({
        'area-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'aria-allowed-attr': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-allowed-role': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-command-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-dialog-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-hidden-body': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-hidden-focus': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-input-field-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-meter-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-progressbar-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-required-attr': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-required-children': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-required-parent': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-roledescription': { priority: 'P3', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-roles': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-toggle-field-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-tooltip-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-treeitem-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-valid-attr': { priority: 'P3', wcagSC: '4.1.2', wcagLevel: 'A' },
        'aria-valid-attr-value': { priority: 'P3', wcagSC: '4.1.2', wcagLevel: 'A' },
        'audio-caption': { priority: 'P1', wcagSC: '1.2.2', wcagLevel: 'A' },
        'autocomplete-valid': { priority: 'P2', wcagSC: '1.3.5', wcagLevel: 'AA' },
        'avoid-inline-spacing': { priority: 'P3', wcagSC: '1.4.12', wcagLevel: 'AA' },
        'blink': { priority: 'P1', wcagSC: '2.2.2', wcagLevel: 'A' },
        'button-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'bypass': { priority: 'P3', wcagSC: '2.4.1', wcagLevel: 'A' },
        'color-contrast': { priority: 'P1', wcagSC: '1.4.3', wcagLevel: 'AA' },
        'definition-list': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'dlitem': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'document-title': { priority: 'P2', wcagSC: '2.4.2', wcagLevel: 'A' },
        'duplicate-id': { priority: 'P3', wcagSC: '4.1.1', wcagLevel: 'A' },
        'duplicate-id-active': { priority: 'P3', wcagSC: '4.1.1', wcagLevel: 'A' },
        'duplicate-id-aria': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'form-field-multiple-labels': { priority: 'P2', wcagSC: '2.5.3', wcagLevel: 'A' },
        'frame-focusable-content': { priority: 'P1', wcagSC: '2.1.1', wcagLevel: 'A' },
        'frame-title': { priority: 'P3', wcagSC: '2.4.2', wcagLevel: 'A' },
        'html-has-lang': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'html-lang-valid': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'html-xml-lang-mismatch': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'image-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'input-button-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'input-image-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'label': { priority: 'P1', wcagSC: '3.3.2', wcagLevel: 'A' },
        'label-title-only': { priority: 'P1', wcagSC: '3.3.2', wcagLevel: 'A' },
        'link-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'list': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'listitem': { priority: 'P2', wcagSC: '1.3.1', wcagLevel: 'A' },
        'marquee': { priority: 'P1', wcagSC: '2.2.2', wcagLevel: 'A' },
        'meta-refresh': { priority: 'P1', wcagSC: '2.2.1', wcagLevel: 'A' },
        'meta-viewport': { priority: 'P2', wcagSC: '1.4.4', wcagLevel: 'AA' },
        'nested-interactive': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'object-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'presentation-role-conflict': { priority: 'P3', wcagSC: '4.1.1', wcagLevel: 'A' },
        'role-img-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'scope-attr-valid': { priority: 'P1', wcagSC: '1.3.1', wcagLevel: 'A' },
        'scrollable-region-focusable': { priority: 'P1', wcagSC: '2.1.1', wcagLevel: 'A' },
        'select-name': { priority: 'P1', wcagSC: '4.1.2', wcagLevel: 'A' },
        'server-side-image-map': { priority: 'P1', wcagSC: '2.1.1', wcagLevel: 'A' },
        'svg-img-alt': { priority: 'P1', wcagSC: '1.1.1', wcagLevel: 'A' },
        'tabindex': { priority: 'P2', wcagSC: '2.4.3', wcagLevel: 'A' },
        'td-headers-attr': { priority: 'P1', wcagSC: '1.3.1', wcagLevel: 'A' },
        'th-has-data-cells': { priority: 'P1', wcagSC: '1.3.1', wcagLevel: 'A' },
        'valid-lang': { priority: 'P2', wcagSC: '3.1.1', wcagLevel: 'A' },
        'video-caption': { priority: 'P1', wcagSC: '1.2.2', wcagLevel: 'A' },
    })
);

export const base = getA11yConfig(baseRulesInfo);
