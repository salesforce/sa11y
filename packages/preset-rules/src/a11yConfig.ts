/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { RunOptions } from 'axe-core';

/**
 * A11yConfig defines options to run accessibility checks using axe specifying list of rules to test
 */
export interface A11yConfig extends RunOptions {
    runOnly: {
        type: 'rule';
        values: string[];
    };
    resultTypes: ['violations'];
}

/**
 * Returns config to be used in axe.run() with given rules
 *
 * @param rules - List of rules to be used in the config
 * @returns A11yConfig with formatted rules
 */
export function getA11yConfig(rules: string[]): A11yConfig {
    return {
        runOnly: {
            type: 'rule',
            values: rules,
        },
        resultTypes: ['violations'],
    };
}
