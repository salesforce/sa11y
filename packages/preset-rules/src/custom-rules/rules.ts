/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import axe from 'axe-core';
const rulesData = [
    {
        id: 'sa11y-Keyboard',
        metadata: {
            description: 'Element is not keyboard operable',
            help: "The following button element <strong>'+ ele.innerText+ '</strong> missing keyboard operability. To fix add tabindex='0' attribute and  appropriate keyboard event handler.",
            helpUrl: '',
        },
        selector:
            "[role='button']:not(a[href],button,input,select,area[href],textarea,[contentEditable=true],[disabled],details)",
        any: [],
        all: ['sa11y-Keyboard-check'],
        none: [],
        tags: ['wcag22aa', 'wcag211'],
    },
];

export default rulesData as axe.Rule[];
