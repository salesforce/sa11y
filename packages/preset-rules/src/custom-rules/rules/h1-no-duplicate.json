{
    "id": "h1-no-duplicate",
    "selector": "h1:not([role]), [role=heading][aria-level=1]",
    "tags": ["cat.semantics", "best-practice"],
    "metadata": {
        "description": "Ensures the document has at most one h1 element",
        "help": "Document must not have more than one h1 element"
    },
    "all": [],
    "any": ["page-no-duplicate-h1"],
    "none": []
}
