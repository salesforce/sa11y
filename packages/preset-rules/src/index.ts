/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { defaultPriority, priorities, wcagLevels, getA11yConfig, RuleInfo } from './rules';
export { defaultRuleset, getDefaultRuleset } from './config';
export { extended } from './extended';
export { base } from './base';
export { full, excludedRules } from './full';
export { WcagMetadata } from './wcag';
export { registerCustomRules } from './customRules';
