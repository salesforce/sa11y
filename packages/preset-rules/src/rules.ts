/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { A11yConfig, log } from '@sa11y/common';

export type WcagVersion = '2.0' | '2.1' | undefined;
export const wcagLevels = ['A', 'AA', 'AAA', ''];
export type WcagLevel = (typeof wcagLevels)[number];
export const priorities = ['P1', 'P2', 'P3', ''] as const;
export type Priority = (typeof priorities)[number];
export const defaultPriority: Priority = 'P3';
export const defaultWcagVersion: WcagVersion = '2.1';

/**
 * Metadata about rules such as Priority and WCAG SC (overriding values from axe tags)
 */
export type RuleMetadata = {
    priority: Priority;
    wcagSC: string;
    wcagLevel: WcagLevel;
};

/**
 * Map of rule IDs to RuleMetadata for convenient lookups of metadata based on rule id
 */
export type RuleInfo = Map<
    string, // Rule ID
    RuleMetadata
>;

/**
 * Get Priority filter for the default ruleset.
 * When set, only the rules matching the given priority from the default ruleset will be
 * used for the sa11y API and automatic checks.
 */
export function getPriorityFilter(): Priority {
    const priorityEnv = process.env.SA11Y_RULESET_PRIORITY;
    const priority = priorityEnv && priorities.includes(priorityEnv as Priority) ? (priorityEnv as Priority) : '';
    if (priority) log(`Setting Sa11y rules priority to ${priority}`);
    return priority;
}

/**
 * Filter given rules and return those matching given priority or .
 */
export function filterRulesByPriority(rules: RuleInfo, priority: Priority = ''): string[] {
    // TODO (refactor): Could be simplified by filtering and returning a RuleInfo map
    //  and using the rule keys in the calling function
    const ruleIDs: string[] = [];
    const priorityOverride = priority || getPriorityFilter();
    if (!priorityOverride) {
        // if no override is set, return all rules
        ruleIDs.push(...rules.keys());
    } else {
        for (const [ruleID, ruleInfo] of rules.entries()) {
            if (priorityOverride === ruleInfo.priority) ruleIDs.push(ruleID);
        }
    }
    return ruleIDs.sort();
}

/**
 * Returns config to be used in axe.run() with given rules
 * Ref: https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#options-parameter
 * @param rules - List of rules to be used in the config
 * @returns A11yConfig with formatted rules
 */
export function getA11yConfig(rules: RuleInfo): A11yConfig {
    // Note: Making the returned config immutable using Object.freeze() results in
    //  "TypeError: Cannot add property reporter, object is not extensible"
    //  even when no local modifications are made. axe.run() itself seems to be modifying the config object.

    return {
        runOnly: {
            type: 'rule',
            values: filterRulesByPriority(rules),
        },
        // Disable preloading assets as it causes timeouts for audio/video elements
        //  with jest and delays webdriver tests by 2-3x when assets are not found (404)
        //  Ref: https://github.com/dequelabs/axe-core/issues/2528
        preload: false,
        // Types not listed will still show a maximum of one node
        resultTypes: ['violations'],
        reporter: 'no-passes', // return only violation results
    };
}
