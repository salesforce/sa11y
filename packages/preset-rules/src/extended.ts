/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { getA11yConfig } from './a11yConfig';
import { recommended } from './recommended';

const rules = [
    ...recommended.runOnly.values,
    'accesskeys',
    'aria-allowed-role',
    'avoid-inline-spacing',
    'css-orientation-lock',
    'frame-tested',
    'identical-links-same-purpose',
    'label-content-name-mismatch',
    'landmark-banner-is-top-level',
    'landmark-complementary-is-top-level',
    'landmark-contentinfo-is-top-level',
    'landmark-main-is-top-level',
    'landmark-no-duplicate-banner',
    'landmark-no-duplicate-contentinfo',
    'landmark-no-duplicate-main',
    'landmark-one-main',
    'landmark-unique',
    'link-in-text-block',
    'meta-viewport-large',
    'meta-viewport',
    'no-autoplay-audio',
    'object-alt',
    'p-as-heading',
    'region',
];

export const extended = getA11yConfig(rules);
