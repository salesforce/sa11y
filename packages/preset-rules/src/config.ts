/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { A11yConfig } from '@sa11y/common';
import { base } from './base';
import { extended } from './extended';
import { full } from './full';

export const ruleSets = new Map(Object.entries({ base: base, extended: extended, full: full }));

/**
 * Get ruleset from environment variable `SA11Y_RULESET` if defined.
 * Defaults to `base` ruleset.
 */
export function getDefaultRuleset(): A11yConfig {
    const envRuleset = process.env.SA11Y_RULESET;
    return envRuleset ? ruleSets.get(envRuleset) || base : base;
}

export const defaultRuleset = getDefaultRuleset();
