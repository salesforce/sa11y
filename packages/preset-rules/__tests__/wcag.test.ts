/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { WcagMetadata } from '../src';
import { Result } from 'axe-core';
import { defaultWcagVersion } from '../src/rules';
import { extendedRulesInfo } from '../src/extended';
import { expect } from '@jest/globals';

// input tags, expected version, expected level, expected SC
const noErrorCases = [
    [['wcag2a'], '2.0', 'A', WcagMetadata.defaultSC],
    [['wcag21aaa'], '2.1', 'AAA', WcagMetadata.defaultSC],
    [['wcag222', 'wcag21aa'], '2.1', 'AA', '2.2.2'],
];

const errorCases = [
    [[], undefined, '', WcagMetadata.defaultSC],
    [[WcagMetadata.defaultSC], undefined, '', WcagMetadata.defaultSC],
    [['wcag222'], undefined, '', '2.2.2'],
    [['foo', 'bar'], undefined, '', WcagMetadata.defaultSC],
    [['wcag2foo', 'wcag21bar'], undefined, '', WcagMetadata.defaultSC],
];

describe('WCAG Metadata extractor', () => {
    it.each([...noErrorCases, ...errorCases])(
        'should extract WCAG version and level (arg:%p)',
        (tags: string[], wcagVersion: string, wcagLevel: string, successCriteria: string) => {
            const wcag = new WcagMetadata({ tags: tags } as Result);
            expect(wcag.wcagVersion).toBe(wcagVersion);
            expect(wcag.wcagLevel).toBe(wcagLevel);
            expect(wcag.successCriteria).toBe(successCriteria);
        }
    );

    it.each(noErrorCases)(
        'should format WCAG metadata when required data is present (arg:%p)',
        (tags: string[], wcagVersion: string, wcagLevel: string, successCriteria: string) => {
            const wcag = new WcagMetadata({ tags: tags } as Result).toString();
            expect(wcag).toMatchSnapshot();
            expect(wcag).toContain(successCriteria);
        }
    );

    it.each([
        'area-alt', // base rule
        'table-duplicate-name', // extended rule without WCAG SC
    ])('should populate WCAG metadata as expected for rule (arg:%p)', (ruleID) => {
        const wcag = new WcagMetadata({ id: ruleID, tags: [] } as Result);
        const ruleInfo = extendedRulesInfo.get(ruleID);
        expect(wcag.wcagVersion).toBe(defaultWcagVersion);
        expect(wcag.wcagLevel).toBe(ruleInfo.wcagLevel);
        expect(wcag.successCriteria).toBe(ruleInfo.wcagSC);
        expect(wcag.priority).toBe(ruleInfo.priority);
    });
});
