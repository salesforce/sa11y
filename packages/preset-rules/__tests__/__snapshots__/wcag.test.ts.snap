// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WCAG Metadata extractor should format WCAG metadata when required data is present (arg:["wcag2a"]) 1`] = `"SA11Y-best-practice-P3"`;

exports[`WCAG Metadata extractor should format WCAG metadata when required data is present (arg:["wcag21aaa"]) 1`] = `"SA11Y-best-practice-P3"`;

exports[`WCAG Metadata extractor should format WCAG metadata when required data is present (arg:["wcag222", "wcag21aa"]) 1`] = `"SA11Y-WCAG-SC2.2.2-P3"`;
